sklearn/model_selection/_classification_threshold.py

```diff
@@ -0,0 +1,1189 @@
+    ----------
+    target_thresholds : ndarray of shape (thresholds,)
+        The thresholds to use to compute the mean score.
+
+    cv_thresholds : ndarray of shape (n_folds, thresholds_fold)
+        The thresholds used to compute the scores for each fold.
+
+    cv_scores : ndarray of shape (n_folds, thresholds_fold)
+        The scores computed for each threshold for each fold.
+
+    Returns
+    -------
+    mean_score : ndarray of shape (thresholds,)
+        The mean score across all folds for each target threshold.
+    """
+    return np.mean(
+        [
+            np.interp(target_thresholds, split_thresholds, split_score)
+            for split_thresholds, split_score in zip(cv_thresholds, cv_scores)
+        ],
+        axis=0,
+    )
+
+
+class TunedThresholdClassifierCV(BaseThresholdClassifier):
+    """Classifier that post-tunes the decision threshold using cross-validation.
+
+    This estimator post-tunes the decision threshold (cut-off point) that is
+    used for converting posterior probability estimates (i.e. output of
+    `predict_proba`) or decision scores (i.e. output of `decision_function`)
+    into a class label. The tuning is done by optimizing a binary metric,
+    potentially constrained by a another metric.
+
+    Read more in the :ref:`User Guide <TunedThresholdClassifierCV>`.
+
+    .. versionadded:: 1.5
+
+    Parameters
+    ----------
+    estimator : estimator instance
+        The classifier, fitted or not, for which we want to optimize
+        the decision threshold used during `predict`.
+
+    objective_metric : {"max_tpr_at_tnr_constraint", "max_tnr_at_tpr_constraint", \
+            "max_precision_at_recall_constraint, "max_recall_at_precision_constraint"} \
+            , str, dict or callable, default="balanced_accuracy"
+        The objective metric to be optimized. Can be one of:
+
+        * a string associated to a scoring function for binary classification
+          (see model evaluation documentation);
+        * a scorer callable object created with :func:`~sklearn.metrics.make_scorer`;
+        * `"max_tnr_at_tpr_constraint"`: find the decision threshold for a true
+          positive ratio (TPR) of `constraint_value`;
+        * `"max_tpr_at_tnr_constraint"`: find the decision threshold for a true
+          negative ratio (TNR) of `constraint_value`.
+        * `"max_precision_at_recall_constraint"`: find the decision threshold for a
+          recall of `constraint_value`;
+        * `"max_recall_at_precision_constraint"`: find the decision threshold for a
+          precision of `constraint_value`.
+
+    constraint_value : float, default=None
+        The value associated with the `objective_metric` metric for which we
+        want to find the decision threshold when `objective_metric` is either
+        `"max_tnr_at_tpr_constraint"`, `"max_tpr_at_tnr_constraint"`,
+        `"max_precision_at_recall_constraint"`, or
+        `"max_recall_at_precision_constraint"`.
```

Comment:
This is in a sense reinventing what in GridSearch we already have as a callable `refit`. I find the API here quite odd. It's also not making the implementation here clean.

I think it would make more sense to accept a callable here that takes the cv results and gives you what needs to be used, then we have 4 pre-defined callables which users and us internally can use, and then we might consider allowing a shortcut for using those callables.

Reply:
I had a similar concern, assuming that users will need more flexibility, e.g. constraint on different metrics. I also don't find the current 4 strings ideal but I understand that they probably represent 90%+ of the use cases, so I think it's acceptable as a start. We can add the option for a callable later if we receive requests.

Reply:
This is what I think users should do:

```py
clf = TunedThreasholdClassifierCV(
    estimator=LogisticRegression(),
    objective=MaxTprAtTnrConstraint(tnr=0.3),
    ...
)
```

And we expose those callable classes to users, and document clearly how users can write one of their own.

Reply:
I also think it's worth special casing the precision/recall and TNR/TPR tradeoffs with the string / float value combo as:
- it makes the very common cases easy to implement without having to write custom filtering logic,
- the intent of an expression such as `TunedThresholdClassifier(clf, objective_metric=max_precision_at_recall_constraint, constraint_value=0.3)` more easily understandable in code than something like:

```python

def max_precision_at_recall(results_at_thresholds, recall=0.3):
   # complex user written code here
   ...
   return selected_threshold_idx

TunedThresholdClassifier(
    clf, objective_metric=["precision", "recall"], constraint=max_precision_at_recall
)
```

So I am +1 for keeping it this way in this first iteration and explore adding more flexibility to custom filtering logic in a follow-up PR.

Reply:
> metrics=["precision", "recall"],

this doesn't look intuitive. There's 1 metric that is optimized. The other one is just to add a constraint on the optimization problem. Your first proposition was better imo: ``objective_metric=MaxTprAtTnrConstraint(tnr=0.3),``.

Reply:
I personally don't mind either way, but I consider both of them better than what we have now. Historically, we haven't had a good experience passing metric name and then metric parameters to the estimators. We've had things like `Estimator(..., metric=..., p=.., metric_params=...)` and I prefer the `metric` passed to have all that information since that's where that information belongs.

Reply:
Well I'd prefer to have metrics as objects as well, but everytime the discussion came up, there was a strong opposition against it.

Reply:
The reason I'm not happy with going forward with this as is, is that I don't want us to release and then immediately deprecated `constraint_value`, which I think doesn't belong to the estimator and it belongs to the constraint.

Reply:
We have many estimators with coupled parameters. On the opposite we almost never have parameters with a complex type. So in fact this would make it a special case in our API. I also think a motivation for keeping separate parameters is to make parameter search easier.

I don't think we'll need to deprecate ``constraint_value`` right away. We can keep the combination of ``objective_metric=<str> + constraint_value=<value>``, easy to use and for most use cases, and add support for ``objective_metric=callable`` for advanced use cases.